Utilize WordPress PHPUnit Test Library Included By `wp-env`

ObliviousHarmony opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This makes the appropriate changes to move us away from needing `wp-phpunit/wp-phpunit`. In the process, I also found that `Yoast/PHPUnit-Polyfills` was not being fully utilized. One of the changes was to change methods like `setUp()` to `set_up()` to prevent PHP 8 compatibility issues.

In addition, as per #39863, the `phpunit` container does not work on M1 chips. This pull request replaces the usage of the `phpunit` container with direct usage of `vendor/bin/phpunit` from Gutenberg.

Closes #39863.

## Why?
#41780 made it so that there's a version of WordPress' PHPUnit test files available within the `wp-env` container. We should dogfood this implementation and use the test files here.

## How?
This pull request makes a few changes to methods in order to be compatible with `Yoast/PHPUnit-Polyfills`:
```
* `setUp()` -> `set_up()`
* `tearDown()` -> `tear_down()`
* `setUpBeforeClass()` -> `set_up_before_class()`
* `setUpAfterClass()` -> `set_up_after_class()`
```

This also removes the `wp-phpunit/wp-phpunit` dependency since it's no longer necessary.

We also add the `wptests_` database table prefix as we are running the test suite in the tests-wordpress container rather than the phpunit container. 

## Testing Instructions
1. Make sure CI passes because it's using the same test script.
2. Run `npm run test-unit-php` and verify that it passes.
3. Run `npm run test-unit-php-multisite` and verify that it passes.
4. Play around with different WordPress versions (it looks like we're not compatible as far back as we claim!)

ObliviousHarmony committed:
Removed `wp-phpunit` Dependency

We're going to be running PHPUnit tests
using the in-container test files. We will
no longer need `wp-phpunit/wp-phpunit`.

ObliviousHarmony committed:
Utilized `Yoast/PHPUnit-Polyfills` In Tests

In order to work on PHP 8, we must use the methods
for setUp(), tearDown(), setUpBeforeClass(), tearDownAfterClass().
This commit replaces those with polyfill versions.

ObliviousHarmony committed:
Stopped Using `phpunit` Container For Tests

This commit replaces the `run phpunit` unit test scripts
with ones that run using the project's PHPUnit.

ObliviousHarmony committed:
Downgraded PHPUnit

ObliviousHarmony committed:
Fixed Multisite Tests

ObliviousHarmony commented on composer.json:
```diff
@@ -31,8 +31,7 @@
 		"sirbrillig/phpcs-variable-analysis": "^2.8",
-		"wp-phpunit/wp-phpunit": "^5.4",
-		"phpunit/phpunit": "^8.5",
+		"phpunit/phpunit": "^6.5",
```
I chose 6.5 here because of the stated requirements for the minimum PHP version. I wanted a version that was compatible with more versions of PHP.

ObliviousHarmony committed:
Removed `wp-phpunit` Constant

noahtallen commented on package.json:
```diff
@@ -297,9 +297,9 @@
 		"test-unit:watch": "npm run test-unit -- --watch",
 		"pretest-unit-php": "wp-env start",
-		"test-unit-php": "wp-env run phpunit \"phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist --verbose\"",
+		"test-unit-php": "wp-env run tests-wordpress /var/www/html/wp-content/plugins/gutenberg/vendor/bin/phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist \"--verbose\"",
```
The quotes around verbose -- do those need to be around all of the arguments passed in? Or are they making wp-env pass them into the container?

ObliviousHarmony replied:
I don't actually know why that was there, I've removed the quotes and it works the same as it did before.

noahtallen replied:
Ah, it must have been left over from when we had to wrap the entire command with quotation marks.

noahtallen commented on phpunit.xml.dist:
```diff
@@ -5,6 +5,7 @@
 	convertErrorsToExceptions="true"
 	convertNoticesToExceptions="true"
 	convertWarningsToExceptions="true"
+	beStrictAboutTestsThatDoNotTestAnything="false"
```
Is there any way to be less heavy-handed? I think this rule could still be useful in general, but I agree we shouldn't warn about the abstract methods.

ObliviousHarmony replied:
I've gone ahead and just marked these tests as incomplete.

noahtallen commented on phpunit/multisite.xml:
```diff
@@ -17,7 +22,6 @@
 	<php>
-		<env name="WP_PHPUNIT__TESTS_CONFIG" value="/wordpress-phpunit/wp-tests-config.php" />
-		<env name="WP_TESTS_DIR" value="" force="true" />
+		<env name="WORDPRESS_TABLE_PREFIX" value="wptests_" />
```
This is duplicated from line 12

ObliviousHarmony replied:
I really should review all of the files after crazy merges, eh? Thanks!

noahtallen replied:
hahaha

ObliviousHarmony committed:
Marked Risky Tests Incomplete

Rather than being heavy-handed and disabling the
strict handling of tests that do nothing, we should
mark those tests as incomplete.

ObliviousHarmony committed:
Removed unnecessary unit test script quoting
